modules/core/04-channel/types/packet.go

```diff
@@ -2,22 +2,42 @@ package types
 
 import (
 	"crypto/sha256"
+	"reflect"
 
 	errorsmod "cosmossdk.io/errors"
 
-	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 
 	clienttypes "github.com/cosmos/ibc-go/v9/modules/core/02-client/types"
 	host "github.com/cosmos/ibc-go/v9/modules/core/24-host"
 	"github.com/cosmos/ibc-go/v9/modules/core/exported"
 )
 
-// CommitPacket returns the packet commitment bytes. The commitment consists of:
+// CommitPacket returns the packet commitment bytes based on
+// the ProtocolVersion specified in the Packet. The commitment
+// must commit to all fields in the packet apart from the source port
+// source channel and sequence (which will be committed to in the packet commitment key path)
+// and the ProtocolVersion which is defining how to hash the packet fields.
+// NOTE: The commitment MUSTs be a fixed length preimage to prevent relayers from being able
+// to malleate the packet fields and create a commitment hash that matches the original packet.
+func CommitPacket(packet Packet) []byte {
+	switch packet.ProtocolVersion {
```

Comment:
if we aren't proving the protocol version a relayer could provide the opposite version that a packet was sent on

Reply:
It would only matter if there was a hash collision between classic packets and eureka packets. Since we changed the commitment algorithm, I don't think that's possible here.

The same exact packet fields with the only thing changed being the ProtocolVersion would lead to different commitments, since the commitment algorithm is different even for the existing fields (and we newly hash dest identifiers)

And additionally two different packets hashing to the same hash by swapping out the ProtocolVersion would break the sha256 preimage collision so we don't need to worry about that.

I have thought about the issue of future maintainer adding a third version. And this third version commits packets in the exact same way as v2. In this case, if the future maintainer is not careful, they can allow relayers to modify the version on receive.

I could signal the possibility of that footgun here by explicitly hashing protocol version. But that would only be a hint for right behavior, since future maintainers may not do it (if they don't change commitment hashes in between the next versions, they will need to hash in the protocol version)